File: MPI/Ring/ex5.c
#include <stdio.h>
#include <mpi.h>

// Rotate values clockwise on a ring with MPI
// ------------------------------------------

int main (int argc, char * argv [])
  {
    
  // Each task owns its private set of the following variables
  // ---------------------------------------------------------     

  int        my_rank, tasks_number;
  char       machine_name [512];
  int        name_length;
  int        my_var, left_neighbour, right_neighbour;
  MPI_Status status;

  
  // We should start with...
  // -----------------------

  MPI_Init (& argc, & argv);


  // Getting information about the MPI environnement
  // -----------------------------------------------

  MPI_Comm_size (MPI_COMM_WORLD, & tasks_number);
  MPI_Comm_rank (MPI_COMM_WORLD, & my_rank);
  MPI_Get_processor_name (machine_name, & name_length);


  // Initialization
  // --------------
  
  my_var = my_rank * 10;

  right_neighbour = (my_rank + 1) % tasks_number;
  left_neighbour =
    (my_rank-1 >= 0)
      ? my_rank-1
      : tasks_number-1; /* rank of the sender */


  // Show initial status
  // -------------------
  
  printf (
    "Hello, my rank is %d among %d tasks on machine %s, with my_var = %d initially\n",
    my_rank, tasks_number, machine_name, my_var
    );
  fflush (stdout);


  // Let's wait until all the tasks have reached this point
  // ------------------------------------------------------

  MPI_Barrier (MPI_COMM_WORLD);

  if (my_rank == 0){
    printf ("--- BARRIER! ---\n");
    fflush (stdout);
    }
    
  
  // Exchange values with the neighbours
  // -----------------------------------

  printf (
    "%s: rank %d sends to %d and receives from %d, my_var = %d\n",
    machine_name, my_rank, right_neighbour, left_neighbour, my_var );

  MPI_Sendrecv_replace (
    & my_var, 1, MPI_INT,
    right_neighbour, 0, left_neighbour, 0,
    MPI_COMM_WORLD, & status );


  // Let's wait until all the tasks have reached this point
  // ------------------------------------------------------

  MPI_Barrier (MPI_COMM_WORLD);

  if (my_rank == 0){
    printf ("--- BARRIER! ---\n");
    fflush (stdout);
    }
    

  // Show final status
  // -------------------
  
  printf (
    "Hello, my rank is %d among %d tasks on machine %s, with my_var = %d finally\n",
    my_rank, tasks_number, machine_name, my_var
    );
  fflush (stdout);


  // ...and we should terminate by...
  // --------------------------------

  MPI_Finalize ();
  
  return 0;
  }

File: MPI/Ring/ex5.f90
PROGRAM main

! Rotate values clockwise on a ring with MPI
! ------------------------------------------

IMPLICIT NONE
INCLUDE 'mpif.h'


! Each task owns its private set of the following variables
! ---------------------------------------------------------

INTEGER       :: my_rank, tasks_number
CHARACTER*32  :: machine_name
INTEGER       :: ierr, name_length
INTEGER	      :: my_var, left_neighbour, right_neighbour
INTEGER       :: status (MPI_STATUS_SIZE)


! We should start with...
! -----------------------

CALL MPI_INIT (ierr)


! Getting information about the MPI environnement
! -----------------------------------------------

CALL MPI_COMM_RANK (MPI_COMM_WORLD, my_rank, ierr)
CALL MPI_COMM_SIZE (MPI_COMM_WORLD, tasks_number, ierr)
CALL MPI_GET_PROCESSOR_NAME (machine_name, name_length, ierr)


! Initialization
! --------------

my_var = my_rank * 10

right_neighbour = MODULO (my_rank+1, tasks_number)
IF (my_rank-1 .GE. 0) THEN
  left_neighbour = my_rank-1
ELSE
  left_neighbour = tasks_number-1 ! rank of the sender
ENDIF


! Show initial status
! -------------------

write ( *, '(a,i5,a,i5,a,a,a,i5,a)' ) &
  & 'Hello, my rank is ', &
  & my_rank, ' among ', tasks_number, &
  & ' tasks on machine ', machine_name (1:name_length), &
  & " with my_var = ", &
  & my_var, &
  & " initially"


! Let's wait until all the tasks have reached this point
! ------------------------------------------------------

CALL MPI_BARRIER (MPI_COMM_WORLD, ierr)

IF( my_rank .EQ. 0 ) THEN
  WRITE ( *, '(a)' ) &
  & "--- BARRIER! ---"
END IF


! Exchange values with the neighbours
! -----------------------------------

CALL MPI_SENDRECV_REPLACE ( &
  & my_var, 1, MPI_INTEGER, &
  & right_neighbour, 0, left_neighbour, 0, &
  & MPI_COMM_WORLD, status, ierr)

WRITE ( *, '(a, a, i2, a, i2, a, i2, a, i2)' ) &
  & machine_name, &
  & ": rank ", &
  & my_var, &
  & " sends to ", &
  & right_neighbour, &
  & " and receives from = ", &
  & left_neighbour, &
  & ", my_var = ", &
  & my_var


! Let's wait until all the tasks have reached this point
! ------------------------------------------------------

CALL MPI_BARRIER (MPI_COMM_WORLD, ierr)

IF( my_rank .EQ. 0 ) THEN
  WRITE ( *, '(a)' ) &
  & "--- BARRIER! ---"
END IF


! Show final status
! -----------------

write ( *, '(a,i5,a,i5,a,a,a,i5,a)' ) &
  & 'Hello, my rank is ', &
  & my_rank, ' among ', tasks_number, &
  & ' tasks on machine ', machine_name (1:name_length), &
  & " with my_var = ", &
  & my_var, &
  & " finally"


! ...and we should terminate by...
! --------------------------------

CALL MPI_FINALIZE (ierr)
  
END PROGRAM main
